sim/battle.ts

```diff
@@ -3105,41 +3107,47 @@ export class Battle extends Dex.ModdedDex {
 		if (this.hints.has(hint)) return;
 
 		if (side) {
-			this.add('split');
-			for (const line of [false, 0, 1, true]) {
-				if (line === true || line === side.n % 2) {
-					this.add('-hint', hint);
-				} else {
-					this.log.push('');
-				}
-			}
+			this.addSplit(side.id, ['-hint', hint]);
 		} else {
 			this.add('-hint', hint);
 		}
 
 		if (once) this.hints.add(hint);
 	}
 
-	add(...parts: (string | number | boolean | ((side: 0 | 1 | boolean) => string) | AnyObject | null | undefined)[]) {
-		if (!parts.some(part => typeof part === 'function')) {
-			this.log.push(`|${parts.join('|')}`);
-			return;
+	addSplit(side: SideID, secret: Part[], shared?: Part[]) {
+		this.log.push(`|split|${side}`);
+		this.add(secret);
+		if (shared) {
+			this.add(shared);
+		} else {
+			this.log.push('');
 		}
-		if (this.reportExactHP) {
-			parts = parts.map(part =>
-				typeof part !== 'function' ? part : part(true)
-			);
+	}
+
+	// FIXME: Using a function to distinguish from AnyObject
+	add(...parts: (Part | (() => {side: SideID, secret: string, shared: string}))[]) {
+		if (!parts.some(part => typeof part === 'function')) {
 			this.log.push(`|${parts.join('|')}`);
 			return;
 		}
-		this.log.push('|split');
-		const sides: (0 | 1 | boolean)[] = [false, 0, 1, true];
-		for (const side of sides) {
-			const sideUpdate = '|' + parts.map(part =>
-				typeof part !== 'function' ? part : part(side)
-			).join('|');
-			this.log.push(sideUpdate);
+		let side: SideID | null = null;
+		const secret = [];
+		const shared = [];
+		for (const part of parts) {
+			if (typeof part === 'function') {
+				const split = part();
+				// FIXME: Can only handle a single side
```

Comment:
I don't think this needs to be fixed, I think this is an intentional design.

Reply:
Unfortunately it doesn't work: data/moves.js:11599 (Pain Split) does two `getHealth` calls, which means we need multiple `split`s I think?

Reply:
So I came up with something which I think works. Before:

```
  '|split',
  '|-sethp|p1a: Gourgeist|44/48|p2a: Dusknoir|48/48|[from] move: Pain Split',
  '|-sethp|p1a: Gourgeist|251/271|p2a: Dusknoir|48/48|[from] move: Pain Split',
  '|-sethp|p1a: Gourgeist|44/48|p2a: Dusknoir|231/231|[from] move: Pain Split',
  '|-sethp|p1a: Gourgeist|251/271|p2a: Dusknoir|231/231|[from] move: Pain Split',
```

after:

```
  '|split|p1',
  '|-sethp|p1a: Gourgeist|251/271|p2a: Dusknoir|[from] move: Pain Split',
  '|-sethp|p1a: Gourgeist|44/48|p2a: Dusknoir|48/48|[from] move: Pain Split',
  '|split|p2',
  '|-sethp|p1a: Gourgeist|44/48|p2a: Dusknoir|231/231|[from] move: Pain Split',
  '|-sethp|p1a: Gourgeist|44/48|p2a: Dusknoir|48/48|[from] move: Pain Split',
```

Its the only rare case where the new protocol is more verbose, but handling it also removes my qualms about `add` being a [partial function](https://wiki.haskell.org/Partial_functions).

Reply:
I'n confused.

Did you mean:

```
|split|p1
|-sethp|p1a: Gourgeist|251/271|[from] move: Pain Split
|-sethp|p1a: Gourgeist|44/48|[from] move: Pain Split
|split|p2
|-sethp|p2a: Dusknoir|231/231|[from] move: Pain Split
|-sethp|p2a: Dusknoir|48/48|[from] move: Pain Split
```

Because that's what I would have done.

Reply:
I did think of this as a possible problem, but when I checked Leech Seed and saw that it didn't do this, I didn't think of Pain Split.

Reply:
Also: `add` is _still_ a partial function. :p

Just throw if you get privileged data from more than one side in a line.

Reply:
> Because that's what I would have done.

`|-sethp|p1a: Gourgeist|251/271|[from] move: Pain Split` alters the definition of `-sethp` and requires client changes, no? You changed the number of args. Also, just noticed the p1 split in my `example` above is wrong, I copied it before I fixed the code.

> Also: `add` is *still* a partial function. :p

Is it? Am I missing something obvious? What valid inputs result in us throwing?

> Just throw if you get privileged data from more than one side in a line.

And modify the Pain Split calls + client?

Reply:
The client is definitely designed to handle single-argument `|-sethp|`. I intentionally designed it for that. The client might need minor updates for animation stuff, and maybe the first one should be tagged `|[silent]`, but that's about it.

By "`add` is a partial function", I mean that it's not valid to pass it two separate splitter functions for different sides, and it's not possible to prevent that with the type system, so it should be prevented by throwing a runtime error.

Reply:
> The client is definitely designed to handle single-argument |-sethp|. I intentionally designed it for that. The client might need minor updates for animation stuff, and maybe the first one should be tagged |[silent], but that's about it.

OK. I didn't document `|-sethp|` (which I will in a followup CL, as well as `|callback|` which is also absent), so I didn't know this. Looking at `src/battle.ts` in the client, youre correct, it looks like it would handle missing a pokemon.

> By "`add` is a partial function", I mean that it's not valid to pass it two separate splitter functions for different sides, and it's not possible to prevent that with the type system, so it should be prevented by throwing a runtime error.

I think my confusion comes from that fact that `add` on HEAD is not a partial function, and my remodified version here is also not partial, but you're saying it *should* be partial, not that it *is* partial. I'm OK with that, I just think we were talking past each other.

Reply:
> (which I will in a followup CL, as well as `|callback|` which is also absent)

In theory `|callback|` should no longer exist...?

> I think my confusion comes from that fact that `add` on HEAD is not a partial function, and my remodified version here is also not partial, but you're saying it _should_ be partial, not that it _is_ partial. I'm OK with that, I just think we were talking past each other.

What's not partial about your remodified version? What happens if you pass it two splitter functions for different sides?

Reply:
> In theory `|callback|` should no longer exist...?

#5414

> What's not partial about your remodified version? What happens if you pass it two splitter functions for different sides?

That's literally what Pain Split is currently doing (passing in two different splitters) and my code now handles it (though, with a different output than you want, as covered above). I'm OK to make `add` partial and change the logic in Pain Split to make two calls though.